File: workout-zio/src/main/scala/workout/domain/WorkoutResponse.scala
package workout.domain

import zio.json._

final case class WorkoutResponse(
    success: Boolean,
    data: Option[Workout],
    message: Option[String]
)

object WorkoutResponse {
  implicit val encoder: JsonEncoder[WorkoutResponse] = DeriveJsonEncoder.gen[WorkoutResponse]
  implicit val decoder: JsonDecoder[WorkoutResponse] = DeriveJsonDecoder.gen[WorkoutResponse]

  def success(workout: Workout): WorkoutResponse =
    WorkoutResponse(success = true, data = Some(workout), message = None)

  def error(msg: String): WorkoutResponse =
    WorkoutResponse(success = false, data = None, message = Some(msg))
}

File: workout-spring-webflux/src/main/java/com/workout/app/api/dto/WorkoutResponse.java
package com.workout.app.api.dto;

import com.workout.app.domain.Workout;

public record WorkoutResponse(boolean success, Workout data, String message) {
    public static WorkoutResponse success(Workout w) {
        return new WorkoutResponse(true, w, null);
    }
    public static WorkoutResponse error(String msg) {
        return new WorkoutResponse(false, null, msg);
    }
}

